File: java/WebFormsCore/Fetch.java
package WebFormsCore;

// WebForms.java 2.0 - The Back-End Part of WebForms Core Technology, Owned by Elanat (https://elanat.net)
// Compatible with WebFormsJS version 2.0

import java.util.Arrays;

/// <summary>
/// Do Not Add Any Data Before Or After It
/// </summary>
public class Fetch {
    // Method
    public static String random(int maxValue) {
        return "@mr" + maxValue;
    }

    public static String random(int minValue, int maxValue) {
        return "@mr" + maxValue + "," + minValue;
    }

    public static String spaceToChar(String text, String character) {
        return "@sc" + character + "," + text;
    }

    public static String spaceToChar(String text) {
        return spaceToChar(text, "-");
    }

    public static String encodeURI(String text) {
        return "@ue" + text;
    }

    public static String decodeURI(String text) {
        return "@ud" + text;
    }

    public static String method(String methodName, String[] args) {
        StringBuilder returnValue = new StringBuilder("@cm" + methodName);

        if (args != null && args.length > 0)
            returnValue.append(",").append(String.join(",", args));

        return returnValue.toString();
    }

    public static String method(String methodName) {
        return method(methodName, null);
    }

    public static String moduleMethod(String methodName, String[] args) {
        StringBuilder returnValue = new StringBuilder("@cM" + methodName);

        if (args != null && args.length > 0)
            returnValue.append(",").append(String.join(",", args));

        return returnValue.toString();
    }

    public static String moduleMethod(String methodName) {
        return moduleMethod(methodName, null);
    }

    public static String wasmMethod(String wasmLanguage, String wasmUrl, String methodName, 
                                  String[] args, String key) {
        StringBuilder returnValue = new StringBuilder("@wA" + wasmLanguage + "," + wasmUrl + "," + methodName);

        if (args != null && args.length > 0)
            returnValue.append(",").append(String.join(",", args));

        return returnValue.toString();
    }

    public static String wasmMethod(String wasmLanguage, String wasmUrl, String methodName, String[] args) {
        return wasmMethod(wasmLanguage, wasmUrl, methodName, args, ".");
    }

    public static String script(String scriptText) {
        return "@_" + scriptText.replace("\n", "$[ln];");
    }

    public static String loadUrl(String url, boolean fetchScript) {
        return "@lu" + url + (fetchScript ? ",1" : "");
    }

    public static String loadUrl(String url) {
        return loadUrl(url, false);
    }

    public static String loadHtml(String url, String fetchInputPlace, boolean fetchScript) {
        return "@lh" + url + "," + (fetchScript ? "1" : "0") + 
               (fetchInputPlace != null && !fetchInputPlace.isEmpty() ? "," + fetchInputPlace : "");
    }

    public static String loadHtml(String url, String fetchInputPlace) {
        return loadHtml(url, fetchInputPlace, false);
    }

    public static String loadLine(String url, int line) {
        return "@ll" + url + "," + line;
    }

    public static String loadINI(String url, String name, boolean isINILike) {
        return "@li" + url + "," + name + (isINILike ? ",1" : "");
    }

    public static String loadINI(String url, String name) {
        return loadINI(url, name, false);
    }

    public static String loadJSON(String url, String name) {
        return "@lj" + url + "," + name;
    }

    public static String loadXML(String url, String name) {
        return "@lx" + url + "," + name;
    }

    public static String hasMethod(String methodName) {
        return "@hm" + methodName;
    }

    public static String hasModuleMethod(String methodName) {
        return "@hM" + methodName;
    }

    public static String getModifierState(String modifier) {
        return "@ms" + modifier;
    }

    // Math
    public static String math(String methodName, String[] args) {
        StringBuilder returnValue = new StringBuilder("@M#" + methodName);

        if (args != null && args.length > 0)
            returnValue.append(",").append(String.join(",", args));

        return returnValue.toString();
    }

    public static String math(String methodName) {
        return math(methodName, null);
    }

    // Data
    public static final String DateYear = "@dy";
    public static final String DateMonth = "@dm";
    public static final String DateDay = "@dd";
    public static final String DateHours = "@dh";
    public static final String DateMinutes = "@di";
    public static final String DateSeconds = "@ds";
    public static final String DateMilliseconds = "@dl";

    // String
    public static final String Space = "@sp";
    public static final String AtSign = "@sa";

    // Tag
    public static String getId(String inputPlace) {
        return "@$i" + inputPlace;
    }

    public static String getName(String inputPlace) {
        return "@$n" + inputPlace;
    }

    public static String getValue(String inputPlace) {
        return "@$v" + inputPlace;
    }

    public static String getValueLength(String inputPlace) {
        return "@$e" + inputPlace;
    }

    public static String getClass(String inputPlace) {
        return "@$c" + inputPlace;
    }

    public static String getStyle(String inputPlace) {
        return "@$s" + inputPlace;
    }

    public static String getTitle(String inputPlace) {
        return "@$l" + inputPlace;
    }

    public static String getLabel(String inputPlace) {
        return "@$A" + inputPlace;
    }

    public static String getText(String inputPlace) {
        return "@$t" + inputPlace;
    }

    public static String getOuterText(String inputPlace) {
        return "@$o" + inputPlace;
    }

    public static String getTextLength(String inputPlace) {
        return "@$g" + inputPlace;
    }

    public static String getAttribute(String inputPlace, String attribute) {
        return "@$a" + inputPlace + "," + attribute;
    }

    public static String getWidth(String inputPlace) {
        return "@$w" + inputPlace;
    }

    public static String getHeight(String inputPlace) {
        return "@$h" + inputPlace;
    }

    public static String getIsReadOnly(String inputPlace) {
        return "@$r" + inputPlace;
    }

    public static String getSelectedIndex(String inputPlace) {
        return "@$x" + inputPlace;
    }

    public static String getIndex(String inputPlace) {
        return "@$I" + inputPlace;
    }

    public static String getTextAlign(String inputPlace) {
        return "@$T" + inputPlace;
    }

    public static String getNodeLength(String inputPlace) {
        return "@$L" + inputPlace;
    }

    public static String getIsVisible(String inputPlace) {
        return "@$V" + inputPlace;
    }

    // Save
    public static String hasHash(String hash) {
        return "@HH" + hash;
    }

    public static String cookie(String key) {
        return "@co" + key;
    }

    public static String session(String key) {
        return "@cs" + key;
    }

    public static String session(String key, String replaceValue) {
        return "@cs" + key + "," + replaceValue;
    }

    public static String sessionAndRemove(String key) {
        return "@cl" + key;
    }

    public static String saved(String key) {
        return session(key);
    }

    public static String saved() {
        return session(".");
    }

    public static String cache(String key) {
        return "@cd" + key;
    }

    public static String cache(String key, String replaceValue) {
        return "@cd" + key + "," + replaceValue;
    }

    public static String cacheAndRemove(String key) {
        return "@ct" + key;
    }

    public static String savedLine(String key, int line) {
        return "@lL" + key + "[" + line;
    }

    public static String savedLine(String key) {
        return savedLine(key, 0);
    }

    public static String savedLineConsume(String key) {
        return "@lL" + key;
    }

    public static String savedINI(String key, String iniKey) {
        return "@lI" + key + "[" + iniKey;
    }

    public static String cacheLine(String key, int line) {
        return "@dL" + key + "[" + line;
    }

    public static String cacheLine(String key) {
        return cacheLine(key, 0);
    }

    public static String cacheLineConsume(String key) {
        return "@dL" + key;
    }

    public static String cacheINI(String key, String iniKey) {
        return "@dI" + key + "[" + iniKey;
    }

    // Format Storage
    public static String formatStore(String key) {
        return "@fr" + key;
    }

    public static String formatStoreByXMLQuery(String key, String xPath) {
        return "@fx" + key + "," + xPath;
    }

    public static String formatStoreByJSONQuery(String key, String query) {
        return "@fj" + key + "," + query;
    }

    public static String formatStoreByINI(String key, String name) {
        return "@fi" + key + "," + name;
    }

    public static String formatStoreByText(String key, int line) {
        return "@ft" + key + "," + line;
    }

    public static String formatStoreByVariable(String key) {
        return "@fv" + key;
    }

    // Document
    public static final String TabIsActive = "@da";

    // Window
    public static final String Href = "@wf";
    public static final String PathName = "@wP";
    public static final String Query = "@wq";
    public static final String Hash = "@wh";
    public static final String Host = "@wH";
    public static final String HostName = "@wn";
    public static final String Port = "@wT";
    public static final String Origin = "@wo";
    public static final String GetSelection = "@ws";
    public static final String ScrollX = "@wx";
    public static final String ScrollY = "@wy";

    // Navigator
    public static final String ClipboardText = "@nC";
    public static final String GeoLatitude = "@nW";
    public static final String GeoLongitude = "@nO";
    public static final String Language = "@nL";
    public static final String IsOnLine = "@no";
    public static final String UserAgent = "@na";

    // Screen
    public static final String ScreenWidth = "@sw";
    public static final String ScreenHeight = "@sh";
    public static final String ScreenOrientationType = "@so";
    public static final String ScreenOrientationAngle = "@sr";

    // Performance
    public static final String TimeOrigin = "@pt";
    public static final String PerformanceNow = "@pn";

    // Event
    public static final String Event = "@EV";
    public static final String EventSerialize = "@Es";
    public static final String EventKey = "@ek";
    public static final String EventWhich = "@ew";
    public static final String EventClientX = "@ex";
    public static final String EventClientY = "@ey";
    public static final String EventPageX = "@eX";
    public static final String EventPageY = "@eY";
    public static final String EventOffsetX = "@Ex";
    public static final String EventOffsetY = "@Ey";
    public static final String EventDeltaY = "@ed";
}

File: elixir/webforms_core/fetch.ex
# webforms.ex 2.1 - The Back-End Part of WebForms Core Technology, Owned by Elanat (https://elanat.net)
# Compatible with WebFormsJS version 2.1

# Do not Add any Data Before or After it
defmodule WebFormsCore.Fetch do
  @rs <<30>>
  @us <<31>>

  # Method
  def random(max_value), do: "@mr" <> to_string(max_value)
  def random(min_value, max_value), do: "@mr" <> to_string(max_value) <> @rs <> to_string(min_value)
  def space_to_char(text, character \\ "-"), do: "@sc" <> character <> @rs <> text
  def encode_uri(text), do: "@ue" <> text
  def decode_uri(text), do: "@ud" <> text

  def method(method_name, args \\ nil) do
    return_value = "@cm" <> method_name

    if args != nil do
      if length(args) > 0 do
        return_value <> @rs <> Enum.map_join(args, @us, &to_string/1)
      else
        return_value
      end
    else
      return_value
    end
  end

  def module_method(method_name, args \\ nil) do
    return_value = "@cM" <> method_name

    if args != nil do
      if length(args) > 0 do
        return_value <> @rs <> Enum.map_join(args, @us, &to_string/1)
      else
        return_value
      end
    else
      return_value
    end
  end

  # MethodName: The Method Name May Need to Include the Class Name, Separated by a Period. Example: MyClassName.MyMethodName
  def wasm_method(wasm_language, wasm_url, method_name, args \\ nil, _key \\ ".") do
    return_value = "@wA" <> wasm_language <> @rs <> wasm_url <> @rs <> method_name

    if args != nil do
      if length(args) > 0 do
        return_value <> @rs <> Enum.map_join(args, @us, &to_string/1)
      else
        return_value
      end
    else
      return_value
    end
  end

  def script(script_text), do: "@_" <> String.replace(script_text, "\n", "$[ln];")
  def load_url(url, fetch_script \\ false), do: "@lu" <> url <> if(fetch_script, do: @rs <> "1", else: "")
  def load_html(url, fetch_input_place \\ "", fetch_script \\ false),
    do: "@lh" <> url <> @rs <> (if fetch_script, do: "1", else: "0") <> if(fetch_input_place in [nil, ""], do: "", else: @rs <> fetch_input_place)
  def load_line(url, line), do: "@ll" <> url <> @rs <> to_string(line)
  def load_ini(url, name, is_ini_like \\ false), do: "@li" <> url <> @rs <> name <> if(is_ini_like, do: @rs <> "1", else: "")
  # Name: Name Or Nested Paths. Is Supprt Index (Student[8].Name). Nested Paths Index Starts At 0
  def load_json(url, name), do: "@lj" <> url <> @rs <> name
  # Name: Name Or XPath; XPath Index Starts At 1
  def load_xml(url, name), do: "@lx" <> url <> @rs <> name
  # MethodName: It's Check Function Or Variable
  def has_method(method_name), do: "@hm" <> method_name
  def has_module_method(method_name), do: "@hM" <> method_name
  # This Method Return True Or False If Key Pressed
  # Modifier: Alt, AltGraph, Control, Meta, Shift, CapsLock, NumLock, ScrollLock
  def get_modifier_state(modifier), do: "@ms" <> modifier

  # Math
  def math(method_name, args \\ nil) do
    return_value = "@M#" <> method_name

    if args != nil do
      if length(args) > 0 do
        return_value <> @rs <> Enum.map_join(args, @us, &to_string/1)
      else
        return_value
      end
    else
      return_value
    end
  end

  # Data
  def date_year, do: "@dy"
  # Month In JavaScript Is Start From Index 0, Month In WebForms Core Is Start From Index 1
  def date_month, do: "@dm"
  def date_day, do: "@dd"
  def date_date, do: "@dD"
  def date_hours, do: "@dh"
  def date_minutes, do: "@di"
  def date_seconds, do: "@ds"
  def date_milliseconds, do: "@dl"

  # String
  def space, do: "@sp"
  def at_sign, do: "@sa"

  # Tag
  def get_id(input_place), do: "@$i" <> input_place
  def get_name(input_place), do: "@$n" <> input_place
  def get_value(input_place), do: "@$v" <> input_place
  def get_value_length(input_place), do: "@$e" <> input_place
  def get_class(input_place), do: "@$c" <> input_place
  def get_style(input_place), do: "@$s" <> input_place
  def get_title(input_place), do: "@$l" <> input_place
  def get_label(input_place), do: "@$A" <> input_place
  def get_text(input_place), do: "@$t" <> input_place
  def get_outer_text(input_place), do: "@$o" <> input_place
  def get_text_length(input_place), do: "@$g" <> input_place
  def get_attribute(input_place, attribute), do: "@$a" <> input_place <> @rs <> attribute
  def get_width(input_place), do: "@$w" <> input_place
  def get_height(input_place), do: "@$h" <> input_place
  def get_is_read_only(input_place), do: "@$r" <> input_place
  def get_selected_index(input_place), do: "@$x" <> input_place
  def get_index(input_place), do: "@$I" <> input_place
  def get_text_align(input_place), do: "@$T" <> input_place
  def get_node_length(input_place), do: "@$L" <> input_place
  def get_is_visible(input_place), do: "@$V" <> input_place

  # Save
  def has_hash(hash), do: "@HH" <> hash
  def cookie(key), do: "@co" <> key
  def save(key \\ "."), do: "@cs" <> key
  def save(key, replace_value), do: "@cs" <> key <> @rs <> replace_value
  def save_then_remove(key), do: "@cl" <> key
  def save_length(key \\ "."), do: "@cg" <> key
  def cache(key \\ "."), do: "@cd" <> key
  def cache(key, replace_value), do: "@cd" <> key <> @rs <> replace_value
  def cache_then_remove(key), do: "@ct" <> key
  def cache_length(key \\ "."), do: "@cG" <> key
  def save_line(key \\ ".", line \\ 0), do: "@lL" <> key <> "[" <> to_string(line)
  def save_line_consume(key \\ "."), do: "@lL" <> key
  # INIKey: Only Direct Key is Supported
  def save_ini(key, ini_key), do: "@lI" <> key <> "[" <> ini_key
  def cache_line(key \\ ".", line \\ 0), do: "@dL" <> key <> "[" <> to_string(line)
  def cache_line_consume(key \\ "."), do: "@dL" <> key
  # INIKey: Only Direct Key is Supported
  def cache_ini(key, ini_key), do: "@dI" <> key <> "[" <> ini_key

  # Format Storage
  def format_store(key), do: "@fr" <> key
  def format_store_by_xml_query(key, xpath), do: "@fx" <> key <> @rs <> xpath
  def format_store_by_json_query(key, query), do: "@fj" <> key <> @rs <> query
  def format_store_by_ini(key, name), do: "@fi" <> key <> @rs <> name
  def format_store_by_text(key, line), do: "@ft" <> key <> @rs <> to_string(line)
  def format_store_by_variable(key), do: "@fv" <> key

  # State
  def has_state(path), do: "@hs" <> path

  # SSE
  def sse_is_connected(path), do: "@Sc" <> path

  # WebSockets
  def web_sockets_is_connected(path \\ ""), do: "@Wc" <> path

  # Document
  def tab_is_active, do: "@da"

  # Window
  def href, do: "@wf"
  def path_name, do: "@wP"
  def query(name \\ "*"), do: "@wq" <> name
  def hash, do: "@wh"
  def host, do: "@wH"
  def host_name, do: "@wn"
  def port, do: "@wT"
  def origin, do: "@wo"
  def get_selection, do: "@ws"
  def scroll_x, do: "@wx"
  def scroll_y, do: "@wy"
  def segment(index), do: "@wS" <> to_string(index)
  # It Only Works when the String Starts with the Tilde Character (~). The Path is Also Separated by the Slash Character (/). #~/Segment1/Segment2/Segment3
  def hash_segment(index), do: "@wt" <> to_string(index)

  # Navigator
  def clipboard_text, do: "@nC"
  def geo_latitude, do: "@nW"
  def geo_longitude, do: "@nO"
  def language, do: "@nL"
  def is_on_line, do: "@no"
  def user_agent, do: "@na"

  # Screen
  def screen_width, do: "@sw"
  def screen_height, do: "@sh"
  def screen_orientation_type, do: "@so"
  def screen_orientation_angle, do: "@sr"

  # Performance
  def time_origin, do: "@pt"
  def performance_now, do: "@pn"

  # Event
  def event, do: "@EV"
  def event_serialize, do: "@Es"
  def event_key, do: "@ek"
  def event_which, do: "@ew"
  def event_client_x, do: "@ex"
  def event_client_y, do: "@ey"
  def event_page_x, do: "@eX"
  def event_page_y, do: "@eY"
  def event_offset_x, do: "@Ex"
  def event_offset_y, do: "@Ey"
  def event_delta_y, do: "@ed"
end